pkg/eventpersisterops/persister/goal_event.go

```diff
@@ -0,0 +1,307 @@
+					Cursor:               cursor,
+				})
+				if err != nil {
+					return nil, err
+				}
+				aor = append(aor, resp.AutoOpsRules...)
+				aorSize := len(resp.AutoOpsRules)
+				if aorSize == 0 || aorSize < listRequestSize {
+					return aor, nil
+				}
+				cursor = resp.Cursor
+			}
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	return exp.([]*aoproto.AutoOpsRule), nil
+}
+
+func (u *evalGoalUpdater) findOpsRules(
+	goalID string,
+	listAutoOpsRules []*aoproto.AutoOpsRule,
+) ([]string, []*goalTargetRule) {
+	featureIDsMap := make(map[string]struct{})
+	targetRules := []*goalTargetRule{}
+	for _, aor := range listAutoOpsRules {
+		autoOpsRule := &aodomain.AutoOpsRule{AutoOpsRule: aor}
+		// We ignore the rules that are already triggered
+		if autoOpsRule.AlreadyTriggered() {
+			continue
+		}
+		clauses, err := autoOpsRule.ExtractOpsEventRateClauses()
+		if err != nil {
+			handledCounter.WithLabelValues(codeFailedToExtractOpsEventRateClauses).Inc()
+			continue
+		}
+		if len(clauses) == 0 {
+			continue
+		}
+		// Find the clauses that contain the goal ID from the the goal event
+		targetClauses := make(map[string]*aoproto.OpsEventRateClause)
+		for id, clause := range clauses {
+			if clause.GoalId == goalID {
+				featureIDsMap[autoOpsRule.FeatureId] = struct{}{}
+				targetClauses[id] = clause
+			}
+		}
+		if len(targetClauses) == 0 {
+			continue
+		}
+		targetRules = append(targetRules, &goalTargetRule{
+			id:        autoOpsRule.Id,
+			featureID: autoOpsRule.FeatureId,
+			clauses:   targetClauses,
+		})
+	}
+	// Convert map to slice
+	featureIDs := make([]string, 0, len(featureIDsMap))
+	for id := range featureIDsMap {
+		featureIDs = append(featureIDs, id)
+	}
+	return featureIDs, targetRules
+}
+
+func (u *evalGoalUpdater) findFeatureVersion(
+	featureID string,
+	features []*featureproto.Feature,
+) int32 {
+	for _, f := range features {
+		if f.Id == featureID {
+			return f.Version
+		}
+	}
+	return 0
+}
+
+func (u *evalGoalUpdater) updateUserCountByRule(
+	environmentNamespace,
+	featureID string,
+	featureVersion int32,
+	userID string,
+	targetRule *goalTargetRule,
+) error {
+	for id, clause := range targetRule.clauses {
+		key := u.newUserCountKey(
+			environmentNamespace,
+			targetRule.id,
+			id,
+			featureID,
+			clause.VariationId,
+			featureVersion,
+		)
+		if err := u.eventCounterCache.UpdateUserCount(key, userID); err != nil {
+			handledCounter.WithLabelValues(codeFailedToUpdateUserCount).Inc()
+			return err
+		}
+		u.logger.Debug(
+			"User count updated successfully",
+			zap.String("pfcountKey", key),
+			zap.String("environmentNamespace", environmentNamespace),
+		)
+	}
+	return nil
+}
+
+func (u *evalGoalUpdater) newUserCountKey(
+	environmentNamespace,
+	ruleID, clauseID, featureID, variationID string,
+	featureVersion int32,
+) string {
+	key := fmt.Sprintf("%s:%s:%s:%d:%s",
+		ruleID,
+		clauseID,
+		featureID,
+		featureVersion,
+		variationID,
+	)
```

Comment:
We need to use pkg/cache/MakeKey() in case the environmentNamespace is "".

Reply:
Do you mean this line?
https://github.com/bucketeer-io/bucketeer/pull/179/files#diff-a1438459208e75caa1a906f3255627b91042a70e54504df4295fa48ab5f81293R306

Reply:
Oh, sorry. I've overlooked 🙏